File: packages/realtime-protocol/src/authorization.ts
export const PROJECT_ROLES = ["viewer", "commenter", "editor", "owner"] as const;
export type ProjectRole = (typeof PROJECT_ROLES)[number];

export const PROJECT_CAPABILITIES = [
  "source_read",
  "presence_join",
  "compile_private",
  "review_read",
  "review_create",
  "review_manage_own",
  "source_write",
  "tree_write",
  "binary_write",
  "review_moderate",
  "suggestion_decide",
  "version_create",
  "version_restore_file",
  "artifact_publish",
  "artifact_select_current",
  "membership_manage",
  "project_controls_manage",
  "review_hide_abuse",
  "version_restore_project",
  "publisher_policy_manage",
  "lifecycle_manage",
  "ownership_transfer",
] as const;

export type ProjectCapability = (typeof PROJECT_CAPABILITIES)[number];

const VIEWER_CAPABILITIES = [
  "source_read",
  "presence_join",
  "compile_private",
  "review_read",
] as const satisfies readonly ProjectCapability[];

const COMMENTER_CAPABILITIES = [
  ...VIEWER_CAPABILITIES,
  "review_create",
  "review_manage_own",
] as const satisfies readonly ProjectCapability[];

const EDITOR_CAPABILITIES = [
  ...COMMENTER_CAPABILITIES,
  "source_write",
  "tree_write",
  "binary_write",
  "review_moderate",
  "suggestion_decide",
  "version_create",
  "version_restore_file",
  "artifact_publish",
  "artifact_select_current",
] as const satisfies readonly ProjectCapability[];

export const ROLE_CAPABILITIES = {
  viewer: VIEWER_CAPABILITIES,
  commenter: COMMENTER_CAPABILITIES,
  editor: EDITOR_CAPABILITIES,
  owner: [
    ...EDITOR_CAPABILITIES,
    "membership_manage",
    "project_controls_manage",
    "review_hide_abuse",
    "version_restore_project",
    "publisher_policy_manage",
    "lifecycle_manage",
    "ownership_transfer",
  ],
} as const satisfies Record<ProjectRole, readonly ProjectCapability[]>;

export function roleHasCapability(
  role: ProjectRole,
  capability: ProjectCapability,
): boolean {
  return (ROLE_CAPABILITIES[role] as readonly ProjectCapability[]).includes(
    capability,
  );
}

File: crates/oleafly-realtime-protocol/src/authorization.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ProjectRole {
    Viewer,
    Commenter,
    Editor,
    Owner,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ProjectCapability {
    SourceRead,
    PresenceJoin,
    CompilePrivate,
    ReviewRead,
    ReviewCreate,
    ReviewManageOwn,
    SourceWrite,
    TreeWrite,
    BinaryWrite,
    ReviewModerate,
    SuggestionDecide,
    VersionCreate,
    VersionRestoreFile,
    ArtifactPublish,
    ArtifactSelectCurrent,
    MembershipManage,
    ProjectControlsManage,
    ReviewHideAbuse,
    VersionRestoreProject,
    PublisherPolicyManage,
    LifecycleManage,
    OwnershipTransfer,
}

const VIEWER: &[ProjectCapability] = &[
    ProjectCapability::SourceRead,
    ProjectCapability::PresenceJoin,
    ProjectCapability::CompilePrivate,
    ProjectCapability::ReviewRead,
];
const COMMENTER: &[ProjectCapability] = &[
    ProjectCapability::SourceRead,
    ProjectCapability::PresenceJoin,
    ProjectCapability::CompilePrivate,
    ProjectCapability::ReviewRead,
    ProjectCapability::ReviewCreate,
    ProjectCapability::ReviewManageOwn,
];
const EDITOR: &[ProjectCapability] = &[
    ProjectCapability::SourceRead,
    ProjectCapability::PresenceJoin,
    ProjectCapability::CompilePrivate,
    ProjectCapability::ReviewRead,
    ProjectCapability::ReviewCreate,
    ProjectCapability::ReviewManageOwn,
    ProjectCapability::SourceWrite,
    ProjectCapability::TreeWrite,
    ProjectCapability::BinaryWrite,
    ProjectCapability::ReviewModerate,
    ProjectCapability::SuggestionDecide,
    ProjectCapability::VersionCreate,
    ProjectCapability::VersionRestoreFile,
    ProjectCapability::ArtifactPublish,
    ProjectCapability::ArtifactSelectCurrent,
];
const OWNER: &[ProjectCapability] = &[
    ProjectCapability::SourceRead,
    ProjectCapability::PresenceJoin,
    ProjectCapability::CompilePrivate,
    ProjectCapability::ReviewRead,
    ProjectCapability::ReviewCreate,
    ProjectCapability::ReviewManageOwn,
    ProjectCapability::SourceWrite,
    ProjectCapability::TreeWrite,
    ProjectCapability::BinaryWrite,
    ProjectCapability::ReviewModerate,
    ProjectCapability::SuggestionDecide,
    ProjectCapability::VersionCreate,
    ProjectCapability::VersionRestoreFile,
    ProjectCapability::ArtifactPublish,
    ProjectCapability::ArtifactSelectCurrent,
    ProjectCapability::MembershipManage,
    ProjectCapability::ProjectControlsManage,
    ProjectCapability::ReviewHideAbuse,
    ProjectCapability::VersionRestoreProject,
    ProjectCapability::PublisherPolicyManage,
    ProjectCapability::LifecycleManage,
    ProjectCapability::OwnershipTransfer,
];

impl ProjectRole {
    pub const fn capabilities(self) -> &'static [ProjectCapability] {
        match self {
            Self::Viewer => VIEWER,
            Self::Commenter => COMMENTER,
            Self::Editor => EDITOR,
            Self::Owner => OWNER,
        }
    }

    pub fn has(self, capability: ProjectCapability) -> bool {
        self.capabilities().contains(&capability)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_editors_and_owners_write_source() {
        assert!(!ProjectRole::Viewer.has(ProjectCapability::SourceWrite));
        assert!(!ProjectRole::Commenter.has(ProjectCapability::SourceWrite));
        assert!(ProjectRole::Editor.has(ProjectCapability::SourceWrite));
        assert!(ProjectRole::Owner.has(ProjectCapability::SourceWrite));
    }
}
